File: src/main/java/com/teamup/teamup/service/NotificationService.java
package com.teamup.teamup.service;

import com.teamup.teamup.dto.NotificationDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Notification service interface.
 */
public interface NotificationService {

    /**
     * Returns a paginated list of unread notifications for a user, newest first.
     *
     * @param userId   the notification receiver
     * @param pageable page number, size, and sort
     * @return page of unread notification DTOs
     */
    Page<NotificationDto> getUnreadNotifications(Long userId, Pageable pageable);

    /**
     * Returns a paginated list of ALL notifications (read + unread) for a user.
     *
     * @param userId   the notification receiver
     * @param pageable page number, size, and sort
     * @return page of all notification DTOs
     */
    Page<NotificationDto> getAllNotifications(Long userId, Pageable pageable);

    /**
     * Returns the count of unread notifications for the badge counter.
     *
     * @param userId the notification receiver
     * @return unread count
     */
    long countUnread(Long userId);

    /**
     * Marks a single notification as read.
     *
     * @param notificationId the notification to mark
     * @param userId        the requesting user (must be the owner)
     * @throws com.teamup.teamup.exception.ResourceNotFoundException if not found
     * @throws java.security.AccessDeniedException                  if userId does not own the notification
     */
    void markAsRead(Long notificationId, Long userId);

    /**
     * Marks all unread notifications as read for a user.
     *
     * @param userId the notification receiver
     * @return number of notifications marked as read
     */
    int markAllAsRead(Long userId);
}

File: frontend/src/services/notification.service.ts
import { apiClient } from '@/lib/axiosInstance'
import type {
  ApiResponse,
  NotificationDto,
  Page,
} from '@/types'

export const notificationService = {
  /**
   * GET /api/notifications
   * Paginated unread notifications, newest first.
   */
  getUnread: (page = 0, size = 20) =>
    apiClient.get<ApiResponse<Page<NotificationDto>>>('/notifications', {
      params: { page, size },
    }),

  /**
   * GET /api/notifications/all
   * Paginated ALL notifications (read + unread), newest first.
   */
  getAll: (page = 0, size = 20) =>
    apiClient.get<ApiResponse<Page<NotificationDto>>>('/notifications/all', {
      params: { page, size },
    }),

  /**
   * GET /api/notifications/unread-count
   * Returns { unreadCount: number }.
   */
  getUnreadCount: () =>
    apiClient.get<ApiResponse<{ unreadCount: number }>>('/notifications/unread-count'),

  /**
   * PUT /api/notifications/{id}/read
   */
  markAsRead: (id: number) =>
    apiClient.put(`/notifications/${id}/read`),

  /**
   * PUT /api/notifications/read-all
   * Returns { markedAsRead: number }.
   */
  markAllAsRead: () =>
    apiClient.put<ApiResponse<{ markedAsRead: number }>>('/notifications/read-all'),
}
